risc0/zkvm/src/prove/dev_mode.rs

```diff
@@ -0,0 +1,89 @@
+// Copyright 2023 RISC Zero, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use alloc::collections::BTreeMap;
+
+use anyhow::Result;
+use risc0_binfmt::MemoryImage;
+
+use crate::{
+    prove::Prover,
+    receipt::{InnerReceipt, Receipt, VerifierContext},
+    Executor, ExecutorEnv, Segment, SegmentReceipt, Session,
+};
+
+/// An implementation of a [Prover] for development and testing purposes.
+pub struct DevModeProver {
+    name: String,
+}
+
+impl DevModeProver {
+    /// Construct a [DevModeProver] with the given name
+    pub fn new(name: &str) -> Self {
+        Self {
+            name: name.to_string(),
+        }
+    }
+}
+
+impl Prover for DevModeProver {
+    fn get_name(&self) -> String {
+        self.name.clone()
+    }
+
+    fn get_peak_memory_usage(&self) -> usize {
+        // note, we don't actually run the prover, so very little memory is consumed
+        // while generating the receipt
+        0
+    }
+
+    fn prove_session(&self, _ctx: &VerifierContext, session: &Session) -> Result<Receipt> {
+        log::info!("prove_session: {}", self.name);
+        eprintln!(
+            "WARNING: Proving in DevMode does not generate a valid receipt. \
```

Comment:
I like `dev mode` or `Dev Mode` because `DevMode` looks like a variable name and we already have an env var and a feature flag thrown into the mix. Gently leaning towards `dev mode` for that reason.

Reply:
I'll go with dev mode for now!